community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/TransactionStream.java

```diff
@@ -0,0 +1,341 @@
+
+package org.neo4j.kernel.impl.store.counts;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.io.fs.StoreChannel;
+import org.neo4j.kernel.impl.transaction.CommittedTransactionRepresentation;
+import org.neo4j.kernel.impl.transaction.command.Command;
+import org.neo4j.kernel.impl.transaction.log.IOCursor;
+import org.neo4j.kernel.impl.transaction.log.LogDeserializer;
+import org.neo4j.kernel.impl.transaction.log.PhysicalLogFile;
+import org.neo4j.kernel.impl.transaction.log.PhysicalLogVersionedStoreChannel;
+import org.neo4j.kernel.impl.transaction.log.PhysicalTransactionRepresentation;
+import org.neo4j.kernel.impl.transaction.log.ReadAheadLogChannel;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntry;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntryCommand;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntryCommit;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntryStart;
+import org.neo4j.kernel.impl.transaction.log.entry.LogHeader;
+
+import static javax.transaction.xa.Xid.MAXBQUALSIZE;
+import static javax.transaction.xa.Xid.MAXGTRIDSIZE;
+
+import static org.neo4j.kernel.impl.transaction.log.LogVersionBridge.NO_MORE_CHANNELS;
+import static org.neo4j.kernel.impl.transaction.log.ReadAheadLogChannel.DEFAULT_READ_AHEAD_SIZE;
+import static org.neo4j.kernel.impl.transaction.log.entry.LogHeaderReader.readLogHeader;
+
+class TransactionStream
+{
+    private final LogFile[] files;
+    private final FileSystemAbstraction fs;
+    private final boolean isContiguous;
+    private final LogFile firstFile;
+
+    public TransactionStream( FileSystemAbstraction fs, File path ) throws IOException
+    {
+        this.fs = fs;
+        { // read metadata from all files in the directory
+            ByteBuffer buffer = ByteBuffer.allocateDirect( 9 + MAXGTRIDSIZE + MAXBQUALSIZE * 10 );
+            File[] files = fs.listFiles( path, new TxFileFilter( fs ) );
+            ArrayList<LogFile> logFiles = new ArrayList<>( files.length );
+            for ( File file : files )
+            {
+                LogFile logFile = new LogFile( fs, file, buffer );
+                if ( logFile.lastCommittedTxId > 0 )
+                {
+                    logFiles.add( logFile );
+                }
+            }
+            this.files = logFiles.toArray( new LogFile[logFiles.size()] );
+        }
+        Arrays.sort( files );
+        { // verify the read metadata
+            LogFile first = null, last = null;
+            boolean isContiguous = true;
+            for ( LogFile file : files )
+            {
+                if ( last == null )
+                {
+                    first = file;
+                }
+                else if ( isContiguous && last.lastCommittedTxId < file.header.lastCommittedTxId )
```

Comment:
This is actually checking the order of the files only, and not whether they are contiguous, with order being the result of the sorting and subsequent check of lastCommittedTxId in the header.

Example:
1,3,5 is ordered
1,2,3,4,5 is ordered and contiguous